pageserver/src/tenant/storage_layer/layer.rs

```diff
@@ -0,0 +1,1115 @@
+        // deleted or detached timeline, don't do anything.
+        let Some(timeline) = self.timeline.upgrade() else { return; };
+
+        // to avoid starting a new download while we evict, keep holding on to the
+        // permit. note that we will not close the semaphore when done, because it will
+        // be used by the re-download.
+        let _permit = {
+            let maybe_downloaded = self.inner.get();
+
+            if version != self.version.load(Ordering::Relaxed) {
+                // downloadness-state has advanced, we might no longer be the latest eviction
+                // work; don't do anything.
+                return;
+            }
+
+            // free the DownloadedLayer allocation
+            match maybe_downloaded.map(|mut g| g.take_and_deinit()) {
+                Some((taken, permit)) => {
+                    assert!(matches!(taken, ResidentOrWantedEvicted::WantedEvicted(_)));
+                    permit
+                }
+                None => {
+                    unreachable!("we do the version checking for this exact reason")
+                }
+            }
+        };
+
+        self.access_stats.record_residence_event(
+            LayerResidenceStatus::Evicted,
+            LayerResidenceEventReason::ResidenceChange,
+        );
+
+        drop(self.status.send(Status::Evicted));
+
+        match capture_mtime_and_remove(&self.path) {
+            Ok(local_layer_mtime) => {
+                let duration = SystemTime::now().duration_since(local_layer_mtime);
+                match duration {
+                    Ok(elapsed) => {
+                        timeline
+                            .metrics
+                            .evictions_with_low_residence_duration
+                            .read()
+                            .unwrap()
+                            .observe(elapsed);
+                        tracing::info!(
+                            residence_millis = elapsed.as_millis(),
+                            "evicted layer after known residence period"
+                        );
+                    }
+                    Err(_) => {
+                        tracing::info!("evicted layer after unknown residence period");
+                    }
+                }
+                timeline
+                    .metrics
+                    .resident_physical_size_gauge
+                    .sub(self.desc.file_size);
+            }
+            Err(e) if e.kind() == std::io::ErrorKind::NotFound => {
+                tracing::info!("failed to evict file from disk, it was already gone");
+            }
+            Err(e) => {
+                tracing::warn!("failed to evict file from disk: {e:#}");
+            }
+        }
+    }
+}
+
+fn capture_mtime_and_remove(path: &Path) -> Result<SystemTime, std::io::Error> {
+    // FIXME: we can now initialize the mtime during first get_or_download,
+    // and track that in-memory for the following? does that help?
```

Comment:
Bogus comment, I forgot we already have that in layer access stats, so this is nothing new.